### PR TITLE
DISTMYSQL-141: cluster-osc-slaves API may not return all required slaves

### PR DESCRIPTION
**Pull request by kamil-holubicki**
DISTMYSQL-141: cluster-osc-slaves API may not return all required slaves

https://jira.percona.com/browse/DISTMYSQL-141

The new algorithm implemented:

----
Stage 1: 1st tier servers.

Get up to two 1st tier servers from each DC in the following order:
1. Most busiest IMs
2. Most lagging leaf nodes
Examples:
1. If there are N > 1 IMs in the DC, we will use the 2 busiest ones
(having the highest number of replicas)
2. If there is only 1 IM in the DC, but there are some leaf nodes,
we will use IM + most lagging leaf node
3. If there are no IMs in the DC, but there are leaf nodes, we will use
up to two most lagging leaf nodes

So this stage will collect at most 2 servers per DC

----
Stage 2: 2nd tier servers

Examine all collected 1st tier servers (from Stage 1),
and if they are IMs, get at most two busiest replicas for each server
(2nd tier servers).

So this stage will collect at most 2 replicas per IM.
If we collected 2 IMs per DC in Stage 1, here we will get 4 servers
per DC.

----
Stage 3: 3rd tier servers

Get 2 busiest 3rd tier replicas per DC.
If replicas are leaves we get the replica with a larger lag.

So this stage will collect at most 2 servers per DC

Added new tests 'get-heuristic-lag-multi-dc' and 'which-cluster-ocs-replicas'

### CODE REVIEW

File: go/inst/instance_dao.go [hunk 1]
```diff
@@ -1738,42 +1750,64 @@ func GetClusterOSCReplicas(clusterName string) ([](*Instance), error) {
...
-				if len(replicas) > 0 {
-					result = append(result, replicas[0])
-				}
-			}
+			*/
 		}
 	}
 	{
 		// Get 2 3rd tier replicas, if possible
+		// todo: again, we get most busy. Why not most laging?
```
Comment (sjmudd) [inline go/inst/instance_dao.go:1810]: typo: lagging
- reasoning: you'd need to ask Shlomi
- I think that the IMs with most replicas are most likely to busier and thus affected by replication delay than general slaves, so the approximation while not perfect is not bad. Is it fully correct? Maybe not.

File: go/inst/instance_dao.go [hunk 2]
```diff
@@ -1728,7 +1740,7 @@ func GetClusterOSCReplicas(clusterName string) ([](*Instance), error) {
 		return [](*Instance){}, log.Errorf("Invalid cluster name: %s", clusterName)
 	}
 	{
-		// Pick up to two busiest IMs
+		// Pick up to two busiest IMs from each DC
```
Comment (sjmudd) [inline go/inst/instance_dao.go:1743]: I suspect there is also a flaw in this logic. There's an implicit assumption that all DCs all have IMs and then slaves under them.
- I know in the environments I work in that is generally the case but I've seen directly attached slaves to a master / gr_primary, and usually they are in the same DC as the master / gr_primary.  Therefore this logic may exclude these slaves if there are no IMs in the same DC as the master / gr_primary and thus ignore them, though for similar reasons they may have replication lag.  Thus these boxes should also be considered.
- e.g. we are trying to get leaf-nodes per DC and the "up to 2" most delayed leaf nodes per DC.  Current code approximates this but does not actually give us this information under some circumstances, and that may cause issues if running an OSC and wanting to throttle if delays are seen.

Comment (kamil-holubicki) [reply go/inst/instance_dao.go:1743]: > and usually they are in the same DC as the master
Isn't this case is covered by the logic at the very end:
// Get 2 1st tier leaf replicas, if possible
?

I'm not sure I understand the topology you are describing.
Should I understand it as:
For every DC get IMs (as implemented) and additionally if there are no IMs in some DC, get 2 leaves from this DC?

Comment (sjmudd) [reply go/inst/instance_dao.go:1743]: It's probably best to respond directly to you on these points.

**Comment by kamil-holubicki** [comment]
https://ps80.cd.percona.com/job/mysql-orchestrator-pipeline/27/

**Review by egegunes — APPROVED** [review]

**Review by EvgeniyPatlan — APPROVED** [review]

**Review by adivinho — APPROVED** [review]

**Pull request merged**